metasyn/filehandler.py

```diff
@@ -0,0 +1,191 @@
+import json
+import warnings
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Optional
+
+import polars as pl
+
+_AVAILABLE_FILE_HANDLERS = {}
+
+
+def filehandler(*args):
+    """Register a dataset so that it can be found by name."""
+
+    def _wrap(cls):
+        _AVAILABLE_FILE_HANDLERS[cls.name] = cls
+        return cls
+
+    return _wrap(*args)
+
+class BaseFileHandler(ABC):
+    name = "base"
+    extensions = []
+
+    def __init__(self, metadata, file_name: str):
+        self.metadata = metadata
+        self.file_name = file_name
+
+    def to_dict(self):
+        return {
+            "file_handler_name": self.name,
+            "format_metadata": self.metadata,
+            "file_name": self.file_name,
+        }
+
+    def save(self, handler_file):
+        with open(handler_file, "w", encoding="utf-8") as handle:
+            json.dump(self.to_dict(), handle, indent=4)
+
+    @abstractmethod
+    def _write_synthetic(self, df, fp):
+        ...
+
+    def write_synthetic(self, df, fp = None):
+        if fp is None:
+            fp = self.file_name
+        if Path(fp).is_file():
+            raise FileExistsError(f"File '{fp}' already exists, choose a different name or write "
+                                  "to a different directory.")
+        self._write_synthetic(df, fp)
+
+
+@filehandler
+class SavFileHandler(BaseFileHandler):
+    name = "spss-sav"
+    extensions = [".sav", ".zsav"]
+
+    def read_dataset(self, fp):
+        df, _prs_metadata = self._get_df_metadata(fp)
+        return df
+        # df, prs_metadata = pyreadstat.read_sav(fp, apply_value_formats=True)
+        # return pl.DataFrame(df)
+
+    @classmethod
+    def _get_df_metadata(cls, fp):
+        try:
+            import pyreadstat
+        except ImportError as err:
+            raise ImportError(
+                "Please install pyreadstat to use the .sav/.zsav file handler.") from err
+
+        pandas_df, prs_metadata = pyreadstat.read_sav(fp, apply_value_formats=True)
```

Comment:
metadata can be read using metadata_only=True (see here https://github.com/Roche/pyreadstat?tab=readme-ov-file#reading-only-the-headers). I think this will return an empty dataframe, so 
```suggestion
        _, prs_metadata = pyreadstat.read_sav(fp, metadataonly=True)
```